pennylane/transforms/get_dag_commutation.py

```diff
@@ -0,0 +1,589 @@
+        if target_control and control_target and target_target:
+            return (
+                bool(commutation_map[operation1.is_controlled][position["ctrl"]])
+                and bool(commutation_map["ctrl"][position[operation2.is_controlled]])
+                and bool(
+                    commutation_map[operation1.is_controlled][position[operation2.is_controlled]]
+                )
+            )
+
+    # Case 3: only operation 1 is controlled
+    elif operation1.is_controlled:
+
+        control_target = intersection(operation1.control_wires, operation2.wires)
+        target_target = intersection(operation1.target_wires, operation2.wires)
+
+        # Case 3.1: control and target 1 overlap with target 2
+        if control_target and target_target:
+            return bool(
+                commutation_map[operation1.is_controlled][position[operation2.name]]
+            ) and bool(commutation_map["ctrl"][position[operation2.name]])
+
+        # Case 3.2: control operation 1 overlap with target 2
+        if control_target and not target_target:
+            return bool(commutation_map["ctrl"][position[operation2.name]])
+
+        # Case 3.3: target 1 overlaps with target 2
+        if not control_target and target_target:
+            return bool(commutation_map[operation1.is_controlled][position[operation2.name]])
+
+    # Case 4: only operation 2 is controlled
+    elif operation2.is_controlled:
+        target_control = intersection(operation1.wires, operation2.control_wires)
+        target_target = intersection(operation1.wires, operation2.target_wires)
+
+        # Case 4.1: control and target 2 overlap with target 1
+        if target_control and target_target:
+            return bool(
+                commutation_map[operation1.name][position[operation2.is_controlled]]
+            ) and bool(commutation_map[operation1.name][position[operation2.is_controlled]])
+
+        # Case 4.2: control operation 2 overlap with target 1
+        if target_control and not target_target:
+            return bool(commutation_map[operation1.name][position["ctrl"]])
+
+        # Case 4.3: target 1 overlaps with target 2
+        if not target_control and target_target:
+            return bool(commutation_map[operation1.name][position[operation2.is_controlled]])
+
+    # Case 5: no controlled operations
+    # Case 5.1: no controlled operations we simply check the commutation table
+    return bool(commutation_map[operation1.name][position[operation2.name]])
+
+
+def _merge_no_duplicates(*iterables):
+    """Merge K list without duplicate using python heapq ordered merging
+
+    Args:
+        *iterables: A list of k sorted lists
+
+    Yields:
+        Iterator: List from the merging of the k ones (without duplicates
+    """
+    last = object()
+    for val in heapq.merge(*iterables):
+        if val != last:
+            last = val
+            yield val
```

Comment:
I'm not so familiar with `heapq`, so curious what logic is happening here (and why `last` needs to be an object type?)